integrations: Create incoming webhook for Patreon.

drrosa opened the pull request:
Creates an incoming webhook integration for Patreon. The main use case is getting notifications when new patrons sign up or update their pledge.

Fixes #18321.
Replaces #22561 and #18517.
CZO [thread](https://chat.zulip.org/#narrow/stream/127-integrations/topic/Patreon.20Integration/near/1408666).

**Notes:**

- Test JSON payloads are different from the actual JSON payloads as noted [here](https://www.patreondevelopers.com/t/why-is-test-payload-data-structure-different-from-real-payloads/4357), and [Patreon's official webhooks documentation](https://docs.patreon.com/#webhooks) is outdated. I had to set up a couple of Patreon test accounts to capture the latest JSON payloads, which don't include data for the total amount of pledges per month.
-  As noted in [Patreon's API docs](https://docs.patreon.com/#triggers-v2), pledge events can be triggered by payment status triggers such as `"Declined"`, `"Pending`", `"Refunded`", `"Fraud`", or `"Other`". We have to check the `last_charge_status` and `patron_status` attributes to make sure we only handle events that correspond to actual membership or pledge events by checking for values such as `"Paid"`, `"Deleted`", `"active_patron"`, `"former_patron"`, or `"null`" .

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/2343554/63a74a7b-c9f3-4620-bb05-b8f11c43380e)
![image](https://github.com/zulip/zulip/assets/2343554/1c700375-02a3-44f8-9a45-561931f46390)

- `members_pledge_create`
![members_pledge_create_monthly](https://github.com/zulip/zulip/assets/2343554/5b164a36-9a3a-4914-bf6a-b2b82889cb19)
![members_pledge_create_yearly](https://github.com/zulip/zulip/assets/2343554/f0d25099-2acd-431b-bcf7-cca8b5003928)
![members_pledge_create_paid_post](https://github.com/zulip/zulip/assets/2343554/c5173094-49e5-4294-a2ed-8e2b853beea7)

- `members_pledge_update`
![members_pledge_update_monthly](https://github.com/zulip/zulip/assets/2343554/27779d2c-721b-41f5-bdbe-3f6131078d40)
![members_pledge_update_yearly](https://github.com/zulip/zulip/assets/2343554/79909302-c7e4-4dd7-9fcb-313e8e2dc9c5)
![members_pledge_update_paid_post](https://github.com/zulip/zulip/assets/2343554/ab4864a3-d531-49fa-8248-e1be77f13798)

- `members_pledge_delete`
![members_pledge_delete](https://github.com/zulip/zulip/assets/2343554/c441d952-a162-4560-9af2-b4a16e049e7b)

- `members_create`
![members_create](https://github.com/zulip/zulip/assets/2343554/e4534d53-4578-456a-8198-a997f2ab6823)

- `members_update`
![members_update_active_patron](https://github.com/zulip/zulip/assets/2343554/9d8f2d5e-d6c9-450f-8850-38c3c2fb7724)
![members_update_former_patron](https://github.com/zulip/zulip/assets/2343554/ff08f037-a75f-4ed2-8e61-9e0ae76ff311)

- `members_delete`
![members_delete](https://github.com/zulip/zulip/assets/2343554/3f3af086-a93e-44db-9c39-0958f054b071)


<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability

Communicate decisions, questions, and potential concerns.

- [x] Automated tests verify logic where appropriate.
- [x] Highlights technical choices and bugs encountered.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

drrosa commented:
@zulipbot add "area: integrations" "documentation review" "maintainer review"

alya commented:
Hm, I'm not sure how to parse the membership pledge messages. Are these recurring contributions scheduled for every month vs. every year? If so, maybe something like this?

> Kopi has pledged $1/month. 🎉 
> Kopi has pledged $1/year. 🎉

alya commented:
I think we generally avoid "patron(s)" type of pluralization. We could do something like this:

> Total number of patrons: 17

drrosa commented:
> Hm, I'm not sure how to parse the membership pledge messages. Are these recurring contributions scheduled for every month vs. every year?

Yes, I double-checked and Patreon only offers [recurring memberships](https://support.patreon.com/hc/en-us/articles/204606215-can-i-make-a-one-time-payment). The value for the payment frequency is provided as an integer value representing the number of months. Since Patreon might start offering 3-month or 6-month subscriptions I thought maybe it would be best to use the exact number of months in the phrasing. How about something like this?


> Kopi has pledged $5.00 for a recurring 1-month subscription.  🎉

> Kopi has pledged $60.00 for a recurring 12-month subscription.  🎉

alya commented:
Hmm, I still find those hard to understand. Without the contexts of how exactly Patreon works, I could interpret that to mean the correct thing, or that the subscription will recur for 12 months and then end.

alya commented:
Is it easy to check what other products do, e.g., some integration via Zapier?

drrosa commented:
Oh actually, I just found a `pay_per_name` JSON field that we can use since it takes on the values: "month", "year", or "creation". So we could phrase the notifications something like:

> Kopi has pledged $5.00 per month.

> Kopi has pledged $60.00 per year.

> Kopi has pledged $10.00 per creation.


What do you think?

alya commented:
The first two are good, but this feels weird:

> Kopi has pledged $10.00 per creation.

Maybe we can just do the following in that case?

> Kopi has pledged $10.00.

drrosa commented:
> The first two are good, but this feels weird:
> 
> > Kopi has pledged $10.00 per creation.
> 
> Maybe we can just do the following in that case?
> 
> > Kopi has pledged $10.00.

Sounds good! I've updated the PR + screenshots.

alya commented:
Thanks! @laurynmm are you up for reviewing this next?

neiljp commented:
@alya It seems odd to ignore the 'per creation' part of that option - it doesn't seem to be a fixed pledge:
https://support.patreon.com/hc/en-us/articles/360002137871-How-per-creation-billing-works

Maybe keep the `per creation`, or based on that article maybe `per paid post`?

alya commented:
Ohh, I thought it was referring to when the account was created. "per paid post" sounds great.

drrosa commented:
@neiljp @alya So I did a little bit more testing and found out that `pay_per_name` is actually a user-defined text value when a Patreon account is set to "per creation" billing. So maybe we should just use that text as is instead of "per paid post"?

![image](https://github.com/zulip/zulip/assets/2343554/32b622ed-2d8d-49dc-bb17-75d47b735c99)

alya commented:
Sure, seems reasonable to me.

drrosa commented:
> Thanks! @laurynmm are you up for reviewing this next?

I tested membership/pledge flows with an actual Patreon account and added a few more conditional checks to make sure notifications are posted accurately. This PR is now ready for a review pass.

CC @sbansal1999 any feedback would be great since you reviewed #28267.

laurynmm commented on zerver/webhooks/patreon/view.py:
```diff
@@ -0,0 +1,205 @@
+
+
+def get_last_charge_status(payload: WildValue) -> object:
+    return payload["data"]["attributes"]["last_charge_status"].tame(check_anything)
```
Should this be `.tame(check_none_or(check_string))` instead of checking for anything? Same not for `get_patron_status`.

drrosa replied:
Yes, you are right! I just didn't realize there was a `check_none_or` function I could use. Thanks!

laurynmm reviewed (commented):
@drrosa - Thanks for working on this! One note first is that you should definitely give yourself a [co-author credit in the commit message](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#mentioning-other-contributors).

I had a few notes about the implementation; let me know if you have any questions.

One overall thought I had about the error messages is that if you passed the event name to the body function, you might be able to use a shared function for generating the two types of error strings - one without `patron_status` for the "members:pledge" events and one with it for the "members" events.

drrosa commented on zerver/webhooks/patreon/view.py:
```diff
@@ -0,0 +1,205 @@
+    user_profile: UserProfile,
+    *,
+    payload: JsonBodyPayload[WildValue],
+    user_specified_topic: OptionalUserSpecifiedTopicStr = None,
```
Got it. Thanks for clarifying this!

drrosa commented on zerver/webhooks/patreon/view.py:
```diff
@@ -0,0 +1,205 @@
+    payload: JsonBodyPayload[WildValue],
+    user_specified_topic: OptionalUserSpecifiedTopicStr = None,
+) -> HttpResponse:
+    header_event = validate_extract_webhook_http_header(request, "X_PATREON_EVENT", "Patreon")
```
Done! I think I incorrectly assumed the formatting was supposed to be similar to the one in the following line:
`fixture_to_headers = get_custom_http_headers_from_filename("HTTP_X_PATREON_EVENT")`
I feel like I don't fully understand why "X-Patreon-Event" is the correct formatting. Should it always be like this by convention or the third-party service decides this?

laurynmm replied:
MDN docs on HTTP Headers: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers. The underscore and all caps string seems to potentially be a vestige from pre-2.2 Django and Zulip's set up for webhook fixture testing  (https://github.com/zulip/zulip/issues/14769).

drrosa commented:
@laurynmm I updated the PR. Thanks for the feedback! I'd say the changes you suggested resulted in a nicely improved implementation.

> @drrosa - Thanks for working on this! One note first is that you should definitely give yourself a [co-author credit in the commit message](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#mentioning-other-contributors).

Thanks for pointing this out! I reset the commit since this PR is essentially a complete rewrite of the two previous implementation attempts given they were based on incorrect JSON payloads. The PRs that are now closed were definitely a useful starting point for me so I gave co-author credits to the corresponding authors. 

> One overall thought I had about the error messages is that if you passed the event name to the body function, you might be able to use a shared function for generating the two types of error strings - one without `patron_status` for the "members:pledge" events and one with it for the "members" events.

I agree. Done!

alya commented:
@laurynmm please mark this PR for integration review whenever it's ready from your perspective.

laurynmm commented on zerver/webhooks/patreon/view.py:
```diff
@@ -0,0 +1,202 @@
+    error_msg = f"{event_name}, where last_charge_status={last_charge_status}"
+    if patron_status:
+        error_msg += f", patron_status={patron_status}"
+    return error_msg
```
I think that `last_charge_status` is always included as a parameter, so it doesn't need to be optional. And then you can have the if statement just be a return ...

```python
    if patron_status:
        return f"{event_name}, ..."
    return f"{event_name}, ..."
```

drrosa replied:
The thing is that mypy throws an incompatible type error if `last_charge_status` is not `Optional[str]` since `get_last_charge_status` will return either `str` or `None`. But having the if statement just be a return works for me.

laurynmm replied:
Right, I wasn't thinking about it being `None` when fetched from the payload. That change works for me.

laurynmm reviewed (commented):
@drrosa - These updates look good. Had one small comment on the error string function.

Even with that comment, I have a question/concern about raising `UnsupportedWebhookEventTypeError` for so many cases.

Do we expect these to happen as valid payloads from Patreon? For example is the `declined_patron` status in the "members:delete" event something that would be potentially sent? And similarly for other values of `last_charged_status`, e.g. `Declined`, that are unsuccessful payments?

If so, it might be better to ignore those events, like we do the deprecated events, because logging them as errors on the server won't be useful in that we know don't want to send a message that payload, but know the payload could be sent?

drrosa commented:
> Do we expect these to happen as valid payloads from Patreon? For example is the `declined_patron` status in the "members:delete" event something that would be potentially sent? And similarly for other values of `last_charged_status`, e.g. `Declined`, that are unsuccessful payments?

Yes, these are valid payloads from Patreon.

> If so, it might be better to ignore those events, like we do the deprecated events, because logging them as errors on the server won't be useful in that we know don't want to send a message that payload, but know the payload could be sent?

The tricky thing here is that we don't want to ignore all of these payloads. For example we do care about "members:delete" event types when `last_charged_status` has a value of either `"Paid"` or `null` to monitor membership terminations.

What if we just drop all the `UnsupportedWebhookEventTypeError(generate_error_msg(...))` statements?

laurynmm commented:
> What if we just drop all the `UnsupportedWebhookEventTypeError(generate_error_msg(...))` statements?

Yeah, that's what I was wondering about, whether we should just not raise those errors. Maybe ask in integrations on CZO?

drrosa commented:
> Yeah, that's what I was wondering about, whether we should just not raise those errors. Maybe ask in integrations on CZO?

https://chat.zulip.org/#narrow/stream/127-integrations/topic/Patreon.20Integration/near/1726271

drrosa commented:
@laurynmm I updated the PR as discussed on CZO. I changed the return value of the event handler functions to return `None` instead of raising an error, updated the logic of the view function, and added tests for the ignored payloads.

laurynmm commented on zerver/webhooks/patreon/view.py:
```diff
@@ -0,0 +1,192 @@
+    if event_name is None:
+        # This is nothing to worry about--get_event() returns None
+        # for events that are valid but not yet handled by us.
+        # See IGNORED_EVENTS, for example.
```
Could this comment just be "See IGNORED_EVENTS"?

laurynmm reviewed (commented):
@drrosa - Looks good! I had two thoughts about the code comments in the api_patreon_webhook function. Feel free to mark with the integration review label once you feel like those comments are resolved.

drrosa commented:
@zulipbot remove "documentation review" "maintainer review"
@zulipbot add “integration review”

drrosa commented:
@zulipbot add “integration review”

> @drrosa - Looks good! I had two thoughts about the code comments in the api_patreon_webhook function. Feel free to mark with the integration review label once you feel like those comments are resolved.

Done, thanks!

drrosa commented:
@zulipbot add "integration review"

> @drrosa - Looks good! I had two thoughts about the code comments in the api_patreon_webhook function. Feel free to mark with the integration review label once you feel like those comments are resolved.

Done, thanks!

timabbott commented:
Merged, thanks @drrosa!
